xunit.runner.wpf/Impl/RemoteTestUtil.DiscoverSession.cs

```diff
@@ -0,0 +1,65 @@
+﻿using System;
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.IO;
+using System.IO.Pipes;
+using System.Linq;
+using System.Text;
+using System.Threading;
+using System.Threading.Tasks;
+using System.Windows.Threading;
+using xunit.runner.data;
+using xunit.runner.wpf.ViewModel;
+
+namespace xunit.runner.wpf.Impl
+{
+    internal partial class RemoteTestUtil 
+    {
+        private sealed class DiscoverSession : ITestDiscoverSession
+        {
+            private readonly Task _task;
+            private event EventHandler<TestCaseDataEventArgs> _testDiscovered;
+            private event EventHandler _sessionFinished;
+
+            internal DiscoverSession(Connection connection, Dispatcher dispatcher, CancellationToken cancellationToken)
+            {
+                _task = BackgroundProducer<TestCaseData>.Go(connection, dispatcher, r => r.ReadTestCaseData(), OnDiscovered, cancellationToken);
+            }
+
+            private void OnDiscovered(List<TestCaseData> list)
+            {
+                Debug.Assert(!_task.IsCompleted);
+
+                if (list == null)
+                {
+                    _sessionFinished?.Invoke(this, EventArgs.Empty);
+                    return;
+                }
+
+                foreach (var cur in list)
+                {
+                    _testDiscovered?.Invoke(this, new TestCaseDataEventArgs(cur));
+                }
+            }
+
+            #region ITestRunSession
+
+            Task ITestSession.Task => _task;
+
+            event EventHandler<TestCaseDataEventArgs> ITestDiscoverSession.TestDiscovered
```

Comment:
The habit I have for personal / side projects is the following: 

- Interfaces are always explicitly implemented
- Interface implementation always forwards to non-interface methods where actual work happens

The logic behind this was simply code hygiene.  I'm constantly refactoring my code which includes adding, deleting and renaming methods.  

Deleting methods in particular caused my problems.  I'd delete a method on an interface, clean up the implementors I knew about and move on.  Inevitably I'd miss a few implementors and essentially leave dead code hanging around.  Later, sometimes years later, I'd come back, find a suspiciously named method, start investigating it for awhile only to eventually figure out it was dead code.  Having explicit implementations saves me from this because it force breaks every consumer when I change any aspect of an interface.  

As said before, it's not a practice I'm wedded to.  It's just an old habit.